File: 3.无重复字符的最长子串.py
#
# @lc app=leetcode.cn id=3 lang=python3
#
# [3] 无重复字符的最长子串
#


# @lc code=start
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if len(s) <= 1:
            return len(s)
        ls = len(s)
        left = 0
        right = 1
        d = {}
        d[s[left]] = left
        maxl = [1 for i in range(ls)]
        while right < ls:
            if not s[right] in d.keys():
                maxl[right] = maxl[right - 1] + 1
                d[s[right]] = right
            else:
                while s[right] in d and left <= d[s[right]]:
                    del d[s[left]]
                    left += 1
                maxl[right] = right - left + 1
                d[s[right]] = right
            right += 1
        return max(maxl)

# @lc code=end

File: 3.无重复字符的最长子串.cpp
/*
 * @lc app=leetcode.cn id=3 lang=cpp
 *
 * [3] 无重复字符的最长子串
 */


// @lcpr-template-start
using namespace std;
#include <algorithm>
#include <array>
#include <bitset>
#include <climits>
#include <deque>
#include <functional>
#include <iostream>
#include <list>
#include <queue>
#include <stack>
#include <tuple>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>
// @lcpr-template-end
// @lc code=start
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, int> window;
        int left = 0, right = 0;
        int res = 0;
        while (right < s.size()) {
            window[s[right]]++;
            while (window[s[right]] > 1) {
                window[s[left++]]--;
            }
            right++;

            res = max(res, right - left);
        }
        return res;
    }
};
// @lc code=end
